cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/IRConsistency.qll

```diff
@@ -546,4 +546,26 @@ module InstructionConsistency {
         "' has no associated variable, in function '$@'." and
     irFunc = getInstructionIRFunction(instr, irFuncText)
   }
+
+  query predicate nonBooleanOperand(
+    Instruction instr, string message, OptionalIRFunction irFunc, string irFuncText
+  ) {
+    exists(Instruction unary |
+      unary = instr.(LogicalNotInstruction).getUnary() and
+      not unary.getResultIRType() instanceof IRBooleanType and
+      irFunc = getInstructionIRFunction(instr, irFuncText) and
+      message =
+        "Logical Not instruction " + instr.toString() +
+          " with non-Boolean operand, in function '$@'."
+    )
```

Comment:
I'm not sure to what extent you want to change the IR behavior here, but this probably requires some care with cases like the following (for which you have already added at test):
```cpp
int y = ...;
int x = !y
```

Reply:
Yep, I intend to fix those cases as well 👍

Reply:
For `!y`, I'd suggest doing the same thing we do for C++, which is to first convert the `int` to `bool` via `CompareNE`.

Separate from the consistency issue, I wonder if translating `!y` directly as `CompareEQ`, instead of `CompareNE, LogicalNot`, would give better results somewhere.

Reply:
> Separate from the consistency issue, I wonder if translating !y directly as CompareEQ, instead of CompareNE, LogicalNot, would give better results somewhere.

The problem with this is the translation back to `Expr`s (which we need in particular when [someone](https://github.com/github/codeql/blob/main/cpp/ql/src/Security/CWE/CWE-290/AuthenticationBypass.ql) does dataflow to a condition of an if statement). In that case, the unconverted result expression should be the `CompareNE`, and the converted result expression should be the `LogicalNot`. But if we merge them together as a `CompareEQ` then the unconverted result expression of `!y` would be a `CompareEQ` which is a bit strange.